### DISCUSSION TITLE
Defining and using Cutom Element with Typescript in vueform

### DISCUSSION THREAD

**Discussion by DerDunkleFux** [vueform/vueform#162] [Questions]
I am currently trying to make a customElement to dynamically load an AG Grid in vueform with vue3 and ts enabled, and after following the https://vueform.com/docs/creating-elements and more resources i am at a loss.
### What is working?
I am able to make a simple vue component and load it as
``` //parent.vue
import AgGrid from '../AgGrid.vue' 
<template>
  <AgGrid></AgGrid>
</template>
```
Where my AgGrid.vue is:
``` //AgGrid.vue
<template>
            <div> ag grid under here {{ customProp }} </div>
            <ag-grid-vue v-bind="gridProp" :style="props.propStyle" :class="props.propClass"> </ag-grid-vue> -->
</template>

<script setup lang="ts">
    console.log("ag grid setup")
    
interface gridPropInterface {
    rowData?: any[];
    columnDefs?: any[];
}

interface AgGridProps {
    gridProp?: gridPropInterface;
    propStyle?: string;
    propClass?: string;
    customProp: string;
}

const props = withDefaults(defineProps<AgGridProps>(), {
    gridProp: () => ({
        rowData: [
        { make: "Tesla", model: "Model Y", price: 64950,},
        { make: "Ford", model: "F-Series", price: 33850,},
        { make: "Toyota", model: "Corolla", price: 29600,},
        ],
        columnDefs:[
            { field: "make" },
            { field: "model" },
            { field: "price" },
        ],
    }),
    propStyle: "width: 100%; height: 100%",
    propClass: "ag-theme-quartz",
    customProp: "my custom prop text"
});
</script> 
```
### What is desired?

I want to be able to load the element in a schema object for vueform like this:
``` //parent.vue
<template #empty>
  <Vueform ref="form$" :tabs="tab_schema" id="tab-form" class="scrollable" :schema="content_schema" #empty :display-errors="false">
  </Vueform>
</template>
```
  Where my content_schema has my elements, formatting and:
  ```
column2: {
                  type: 'group',
                  schema: {
                    gridDescription: {
                      type:"static",
                      tag:"h4",
                      content:"ag grid here",
                    },
                    myAgGrid: {
                      type: 'AgGrid',
                      }
                  },
                  size: 'lg',
                  columns: {
                    container: 7,
                  },
              },
```
 But AgGrid is not being recognized
 ### What have i tried?

I registered the element in vueform.config.ts:
 ``` //vueform.config.ts
 import AgGrid from './src/components/AgGrid.vue';

console.log("in config")
console.log(AgGrid)
export default defineConfig({
  theme: vueform,
  locales: { en },
  locale: 'en',
  elements : [
    AgGrid
  ]
});
```
And followed the steps to make my element functional, except for two things that i couldn't do because of typescript:
1.  I could not place the #element inside the `<ElementLayout> <template #element>...` because i get this error:
`Property 'element' does not exist on type '{ field: VNode<RendererNode, RendererElement, { [key: string]: any; }>[]; label: VNode<RendererNode, RendererElement, { ...; }>[]; ... 4 more ...; after: VNode<...>[]; }'.
`
2. If i use defineElement from @vueform/vueform no props are recognized because i get: 
`Property 'customProp' does not exist on type '{}'` 
So i use defineComponent, bc of standard vue definition.

Here is the code that made the most sense but still got errors, i let the ag grid props away a just try to get the customProp to work but it didn't:
``` //AgGrid.vue
<script lang="ts">
// ...imports...
export default defineComponent({
    name: 'AgGrid',
    props: {
        customProp: {
            type:String,
            default:"my custom prop!",
        },
    },
    mounted() {
        console.log("ag grid mounted")
    },
    setup(props:any, { element }:any) {
        console.log("ag grid setup")
    }
     data() {
        return {
            defaultClasses: {
                container: '',
            }
        }
    }
});
 </script>
 <template #element>
    <ElementLayout> 
        <template>
            <div> ag grid under here {{ customProp }}</div>
        </template>
    </ElementLayout>
 </template>
```
So now if i try to load it as <AgGrid></AgGrid> without explicitly importing i get following errors:
` TypeError: this.template is undefined`
BUT i get the console logs for mounted and setup :)

And if i try to load it as a schema object (desired functionality)
`myAgGrid: {
                      type: 'AgGrid',...` nothing at all happens
Im very grateful for any kind of help :)

**Comment by adamberecz** [comment]
The trick is that you need to append `Element` to the element name:

```vue
<script lang="ts">
// ...imports...
export default defineComponent({
    name: 'AgGridElement',
    // ...
});
</script>
 ```
